lib/parsers/ts-utils.js

```diff
@@ -0,0 +1,54 @@
+const fs = require('node:fs');
+const ts = require('typescript');
+const { transformForLint } = require('./transform');
+const babel = require('@babel/core');
+const { replaceRange } = require('./transform');
+
+module.exports.patchTs = function patchTs() {
+  const readDirectory = ts.sys.readDirectory;
+  ts.sys.readDirectory = function (...args) {
+    const results = readDirectory.call(this, ...args);
+    return results.map((f) => f.replace(/\.gts$/, '.mts'));
+  };
+  ts.sys.fileExists = function (fileName) {
+    return fs.existsSync(fileName.replace(/\.mts$/, '.gts')) || fs.existsSync(fileName);
+  };
+  ts.sys.readFile = function (fname) {
+    let fileName = fname;
+    let content = '';
+    try {
+      content = fs.readFileSync(fileName).toString();
+    } catch {
+      fileName = fileName.replace(/\.mts$/, '.gts');
+      content = fs.readFileSync(fileName).toString();
+    }
+    if (fileName.endsWith('.gts')) {
+      content = transformForLint(content).output;
+    }
+    if (fileName.endsWith('.ts') || fileName.endsWith('.gts')) {
```

Comment:
@patricklx I think I found another bug.
Basically I noticed that `replaceExtensions` is erroring for a lot of declaration files coming from `node_modules/@types`. You won't see the errors logged in test here because they are silent errors but they do happen. And it seems that on some cases we break some things which actually make some tests in sonarjs break

For example try to see what happens when it tried to do `replaceExtensions` for `@type/babel__core/index.d.ts` file (it will throw an error)
Right now babel.parse inside `replaceExtensions` isn't even set up to handle `d.ts` files, and even when I add it `dts: true` I'm still seeing some errors. Do you think we can fix this somehow?

maybe make `replaceExtensions` error resilient in case of `.d.ts` files, or make sure it works on all declaration files it tries to read

Reply:
Or maybe just make sure it runs `replaceExtensions` on our own code and not stuff coming from elsewhere?

Reply:
i think we can just ignore .d.ts files

Reply:
normally when plugins with gts files are released, all should already be compiled to .js and .d.ts files

Reply:
If we can just ignore declaration files for this then great :D